install/install.ps1

```diff
@@ -1,94 +1,203 @@
 # Enable a default -Verbose flag for debug output.
 [CmdletBinding()]
-Param()
+Param([switch] $NoUpdatePath)
 
 # Halt script execution at the first failed command.
 $script:ErrorActionPreference='Stop'
 
 # We don't currently have good cross-platform options for determining the latest release version, so we require
 # knowledge of the specific target version, which the user should set as an environment variable.
+
 $version = $Env:AZUREAUTH_VERSION
 if ([string]::IsNullOrEmpty($version)) {
     Write-Error 'No $AZUREAUTH_VERSION specified, unable to download a release'
 }
 
-$repo = if ([string]::IsNullOrEmpty($Env:AZUREAUTH_REPO)) { 'AzureAD/microsoft-authentication-cli' } else { $Env:AZUREAUTH_REPO }
-$releaseName = "azureauth-${version}-win10-x64"
-$releaseFile = "${releaseName}.zip"
-$releaseUrl = "https://github.com/${repo}/releases/download/${version}/$releaseFile"
+function Install-Pre-0-4-0 {
 
-$azureauthDirectory = if ([string]::IsNullOrEmpty($Env:AZUREAUTH_INSTALL_DIRECTORY)) {
-    ([System.IO.Path]::Combine($Env:LOCALAPPDATA, "Programs", "AzureAuth"))
-} else {
-    $Env:AZUREAUTH_INSTALL_DIRECTORY
-}
-$extractedDirectory = ([System.IO.Path]::Combine($azureauthDirectory, $releaseName))
-$latestDirectory = ([System.IO.Path]::Combine($azureauthDirectory, "latest"))
-$zipFile = ([System.IO.Path]::Combine($azureauthDirectory, $releaseFile))
-
-Write-Verbose "Creating ${azureauthDirectory}"
-$null = New-Item -ItemType Directory -Force -Path $azureauthDirectory
-
-Write-Verbose "Downloading ${releaseUrl} to ${zipFile}"
-$client = New-Object System.Net.WebClient
-$client.DownloadFile($releaseUrl, $zipFile)
-
-# A running instance of azureauth can cause installation to fail, so we try to kill any running instances first.
-# We suppress taskkill output here because this is a best effort attempt and we don't want the user to see its output.
-# Here, Get-Process is used to first determine whether there is an existing azureauth process. If there is, kill the existing process first.
-$ProcessCheck = Get-Process -Name azureauth -ErrorAction SilentlyContinue -ErrorVariable ProcessError
-if ($ProcessCheck -ne $null)
-{
-    Write-Verbose "Stopping any currently running azureauth instances"
-    taskkill /f /im azureauth.exe 2>&1 | Out-Null
-
-    # After killing the process it is still possible for there there to be locks on the files it was using (including
-    # its own DLLs). The OS may take an indeterminate amount of time to clean those up, but so far we've observed 1
-    # second to be enough.
-    Start-Sleep -Seconds 1
-}
+    Write-Verbose "Installing using pre-0.4.0 method"

```

Comment:
Will we need to change the version name manually in this script? If yes, how do we remember to do it?

Reply:
This version represents a specific point in time at which we radically changed the installation logic. We shouldn't need to change it when we release new versions. Ideally this goes away as well, we just need to leave some time for users to migrate to a post-0.4.0 installation (maybe 6 months?).

Reply:
Perfect! So there is a world on how the logic works pre 0.4.0 and post 0.4.0.